include/cocaine/detail/engine.hpp

```diff
@@ -59,8 +62,13 @@ class execution_unit_t {
 
    ~execution_unit_t();
 
-    std::shared_ptr<session_t>
-    attach(const std::shared_ptr<asio::ip::tcp::socket>& ptr, const io::dispatch_ptr_t& dispatch);
+    template<class Socket>
+    std::shared_ptr<session<typename Socket::protocol_type>>
+    attach(std::unique_ptr<Socket> ptr, const io::dispatch_ptr_t& dispatch);
+
+    template<class Socket>
+    std::shared_ptr<session<typename Socket::protocol_type>>
+    attach(const std::shared_ptr<Socket>& ptr, const io::dispatch_ptr_t& dispatch);
```

Comment:
Because of https://github.com/cocaine/cocaine-core/blob/v0.12/src/service/locator.cpp#L401